packages/data-flow/test/registries/cachedEventRegistry.spec.ts

```diff
@@ -0,0 +1,107 @@
+
+import { NewProcessedEvent, ProcessedEvent } from "@grants-stack-indexer/repository";
+import { ChainId, ILogger } from "@grants-stack-indexer/shared";
+
+import { IEventsRegistry } from "../../src/internal.js";
+import { InMemoryCachedEventRegistry } from "../../src/registries/event/cachedEventRegistry.js";
+
+describe("InMemoryCachedEventRegistry", () => {
+    const logger: ILogger = {
+        debug: vi.fn(),
+        error: vi.fn(),
+        info: vi.fn(),
+        warn: vi.fn(),
+    };
+
+    const mockEventRegistry: IEventsRegistry = {
+        getLastProcessedEvent: vi.fn(),
+        saveLastProcessedEvent: vi.fn(),
+    };
+
+    const chainId = 1 as ChainId;
+    const mockEvent: ProcessedEvent = {
+        chainId,
+        blockNumber: 100,
+        blockTimestamp: 1234567890,
+        logIndex: 1,
+    };
+
+    beforeEach(() => {
+        vi.clearAllMocks();
+    });
+
+    it("initialize with existing events", async () => {
+        vi.mocked(mockEventRegistry.getLastProcessedEvent).mockResolvedValue(mockEvent);
+
+        const registry = await InMemoryCachedEventRegistry.initialize(logger, mockEventRegistry, [
+            chainId,
+        ]);
+
+        const cached = await registry.getLastProcessedEvent(chainId);
+        expect(cached).toEqual(mockEvent);
+        expect(mockEventRegistry.getLastProcessedEvent).toHaveBeenCalledTimes(1);
+    });
+
+    it("fetch from underlying registry when not in cache", async () => {
+        vi.mocked(mockEventRegistry.getLastProcessedEvent)
+            .mockResolvedValueOnce(undefined) // For initialization
+            .mockResolvedValueOnce(mockEvent); // For actual fetch
+
+        const registry = await InMemoryCachedEventRegistry.initialize(logger, mockEventRegistry, [
+            chainId,
+        ]);
+
+        const result = await registry.getLastProcessedEvent(chainId);
+        expect(result).toEqual(mockEvent);
+        expect(mockEventRegistry.getLastProcessedEvent).toHaveBeenCalledTimes(2);
+    });
+
+    it("save event and update cache", async () => {
+        const registry = await InMemoryCachedEventRegistry.initialize(logger, mockEventRegistry, [
+            chainId,
+        ]);
+
+        const newEvent: NewProcessedEvent = {
+            blockNumber: 200,
+            blockTimestamp: 1234577890,
+            logIndex: 2,
+        };
+
+        await registry.saveLastProcessedEvent(chainId, newEvent);
+
+        // Verify the event was saved to underlying registry
+        expect(mockEventRegistry.saveLastProcessedEvent).toHaveBeenCalledWith(chainId, newEvent);
+
+        // Verify the cache was updated
+        const cached = await registry.getLastProcessedEvent(chainId);
+        expect(cached).toEqual({
+            ...newEvent,
+            chainId,
+        });
+
+        // Verify no additional calls to underlying registry
+        expect(mockEventRegistry.getLastProcessedEvent).toHaveBeenCalledTimes(1);
+    });
+
```

Comment:
What if the other registry fails while saving?

Reply:
that's a really good question sir

cc @0xkenj1 we can discuss it offline if this is is a retriable error or not